From 9e95ac10f73a48ceaa2be037286a9f9aa268b099 Mon Sep 17 00:00:00 2001
From: William Douglas <william.douglas@intel.com>
Date: Thu, 4 Apr 2024 10:56:18 -0700
Subject: [PATCH] Rework how curl errors are checked

In cases where curl perform returns an error but is actually a 400
level response, swupd needs to check curl info to see this. If the
info reveals a 400 level response, return the corresponding download
status error to prevent useless download retries.

Signed-off-by: William Douglas <william.douglas@intel.com>
---
 src/swupd_lib/curl.c | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/src/swupd_lib/curl.c b/src/swupd_lib/curl.c
index 8def18f8..f3577d7c 100644
--- a/src/swupd_lib/curl.c
+++ b/src/swupd_lib/curl.c
@@ -456,6 +456,23 @@ enum download_status process_curl_error_codes(int curl_ret, CURL *curl_handle)
 			return DOWNLOAD_STATUS_ERROR;
 		}
 	} else { /* download failed but let our caller do it */
+		long response = 0;
+		curl_easy_getinfo(curl_handle, CURLINFO_RESPONSE_CODE, &response);
+		debug("Curl - process_curl_error_codes: curl_ret = %d, response = %ld\n", curl_ret, response);
+		switch (response) {
+		case 403:
+			debug("Curl - Download failed - forbidden (403) - '%s'\n", url);
+			return DOWNLOAD_STATUS_FORBIDDEN;
+		case 404:
+			debug("Curl - Download failed - file not found (404) - '%s'\n", url);
+			return DOWNLOAD_STATUS_NOT_FOUND;
+		case 416:
+			debug("Curl - Download failed - range not satisfiable (416) - '%s'\n", url);
+			return DOWNLOAD_STATUS_RANGE_NOT_SATISFIABLE;
+		default:
+			debug("Curl - Download failed: response (%ld) -  '%s'\n", response, url);
+		}
+
 		debug("Curl - process_curl_error_codes - curl_ret = %d\n", curl_ret);
 		switch (curl_ret) {
 		case CURLE_COULDNT_RESOLVE_PROXY:
-- 
2.44.0

